dotnet/src/Connectors/Connectors.Memory.Sqlite/SqliteVectorStoreRecordCollection.cs

```diff
@@ -227,85 +239,131 @@ public Task<VectorSearchResults<TRecord>> VectorizedSearchAsync<TVector>(TVector
+        await foreach (var record in this.InternalGetBatchAsync(connection, keys, options, cancellationToken).ConfigureAwait(false))
+        {
+            yield return record;
+        }
     }
 
     /// <inheritdoc />
-    Task<string> IVectorStoreRecordCollection<string, TRecord>.UpsertAsync(TRecord record, CancellationToken cancellationToken)
+    async Task<string> IVectorStoreRecordCollection<string, TRecord>.UpsertAsync(TRecord record, CancellationToken cancellationToken)
     {
-        return this.InternalUpsertAsync<string>(record, cancellationToken);
+        using var connection = await this.GetConnectionAsync(cancellationToken).ConfigureAwait(false);
+        return await this.InternalUpsertAsync<string>(connection, record, cancellationToken)
+            .ConfigureAwait(false);
     }
 
     /// <inheritdoc />
-    IAsyncEnumerable<string> IVectorStoreRecordCollection<string, TRecord>.UpsertBatchAsync(IEnumerable<TRecord> records, CancellationToken cancellationToken)
+    async IAsyncEnumerable<string> IVectorStoreRecordCollection<string, TRecord>.UpsertBatchAsync(
+        IEnumerable<TRecord> records,
+        [EnumeratorCancellation] CancellationToken cancellationToken)
     {
-        return this.InternalUpsertBatchAsync<string>(records, cancellationToken);
+        using var connection = await this.GetConnectionAsync(cancellationToken).ConfigureAwait(false);
+        await foreach (var record in this.InternalUpsertBatchAsync<string>(connection, records, cancellationToken)
+                           .ConfigureAwait(false))
+        {
+            yield return record;
+        }
     }
 
     /// <inheritdoc />
-    public Task DeleteAsync(string key, CancellationToken cancellationToken = default)
+    public async Task DeleteAsync(string key, CancellationToken cancellationToken = default)
     {
-        return this.InternalDeleteAsync(key, cancellationToken);
+        using var connection = await this.GetConnectionAsync(cancellationToken).ConfigureAwait(false);
+        await this.InternalDeleteAsync(connection, key, cancellationToken)
+            .ConfigureAwait(false);
     }
 
     /// <inheritdoc />
-    public Task DeleteBatchAsync(IEnumerable<string> keys, CancellationToken cancellationToken = default)
+    public async Task DeleteBatchAsync(IEnumerable<string> keys, CancellationToken cancellationToken = default)
     {
-        return this.InternalDeleteBatchAsync(keys, cancellationToken);
+        using var connection = await this.GetConnectionAsync(cancellationToken).ConfigureAwait(false);
+        await this.InternalDeleteBatchAsync(connection, keys, cancellationToken).ConfigureAwait(false);
     }
 
     #endregion
 
     #region private
 
+    private async ValueTask<SqliteConnection> GetConnectionAsync(CancellationToken cancellationToken = default)
+    {
+        var connection = new SqliteConnection(this._connectionString);
```

Comment:
Does `SqliteConnection` type implements it's own private pool of connections? So we don't actually open a new connection every time this method is being called?

Reply:
Yep, that's exactly right - that's the pattern that all .NET database drivers follow in general.

Static/private/implicit pools are (IMHO) generally a bad idea, which is why we introduced the DbDataSource abstraction - that's basically meant to be a factory for connections, and can explicitly represent a connection pool; so it allows users to manage pools explicitly (plus other benefits). Unfortunately, for now only Npgsql properly implements this.